### ISSUE TITLE
All Sorting methods should be public

### ISSUE THREAD

**Issue by Speiger** [vigna/fastutil#172]
FastUtil has a lot of Methods for sorting.
- Insertion Sort is not public and locked behind a ObjectAllocation. So it should be made public.
- Merge sort always copies the array even if it is not nessesary. It should only copy the array if the size is bigger then 16, and only copy the range it is trying to sort.
- Selection sort is private too even so it has its usecases. But it is locked behind Radix sort or quicksort.
- Also #107

**Comment by incaseoftrouble** [comment]
Insertion sort and selection sort have very limited applications (typically only for arrays < 16ish entries) due to their horrible runtime performance.

**Comment by Speiger** [comment]
Still better then using merge sort with less then 16 elements since that is effectivly a arraycopy.

**Comment by incaseoftrouble** [comment]
That's why they are called for small arrays by the asymptotically faster algorithms.

**Comment by Speiger** [comment]
And what has that to do that you are forced to use a memory allocating method to have a small array sorter that you would copy paste to be able to use instead of calling the memory allocating function that calls the function you copy instead?

**Comment by incaseoftrouble** [comment]
1) Indeed, the `mergeSort` call should directly delegate to `insertionSort` before cloning. That's easy to fix but doesn't cost much in any case. I think you overestimate the cost of cloning an array with 10ish entries.
 2) Exposing selection / insertion sort incurs the risk of people actually using it. Please never use these directly. Their only purpose is to be used at the leafs of the merge/quicksort computation tree.
 3) You are not forced to use a memory allocating sorting (quickSort is in-place)

**Comment by Speiger** [comment]
1: My projects are usually  those where micro optimizations are "very" useful to performance of the final project.
2: True, that makes sense with what i have been complaining about. So at least consistent.
3: Yeah.... Like there isn't any reason to replace selection sort with insertion sort, like they are not 2 different sorting algorithms with strengths and weaknesses.

**Comment by incaseoftrouble** [comment]
> 1: My projects are usually those where micro optimizations are "very" useful to performance of the final project.

I'd be happy to see a real life example where this change has noticable impact. I guess You'd have to sort millions of small arrays and do nothing else. But again, as I said, adding that check in the "basic" merge sort call does make sense. I'm only saying that the effects probably won't be severe, but I'd be happy to be wrong!

>  3: Yeah.... Like there isn't any reason to replace selection sort with insertion sort, like they are not 2 different sorting algorithms with strengths and weaknesses.

There is *no* reason to use either of them (except for very special applications, and even then you would really need to know what you are doing. But in that case you would fine-tune your sort implementation anyway; nobody is stopping you from doing that btw).

**Comment by Speiger** [comment]
- 1: Large scale i already have where quicksort helps a lot where i had to sort 2.4 Million Objects, It is so much faster then other implementations so far.
- 3: Should be public exactly for those who want to optimize in the first place. More options are more options.

**Comment by incaseoftrouble** [comment]
* 1: Large scale i already have where quicksort helps a lot where i had to sort 2.4 Million Objects, It is so much faster then other implementations so far.

On arrays larger than 16 this modification doesn't do anything! I don't see what quicksort has to do with that discussion. Nevertheless, mergesort is *MUCH* faster than insertion or selection sort on random arrays. And yes, quicksort is the preferred way to sort arrays (well, introsort / med3 / double pivot quicksort currently), mergesort typically is better for linked data structures.

> 3: Should be public exactly for those who want to optimize in the first place. More options are more options.

More public options are more things to maintain, more potential for misuse etc. The people who want to really optimize their sorting can copy and modify the sources appropriately.

**Comment by vigna** [comment]
Exposing all those sorting methods would just create confusion and pollute the documentation.

Good point reminding me of #107 , which I completely forgot about. I also check the parallelism of the common pool and if it's 1 I switch to the non-parallel implementation.

I fixed the mergesort useless allocation in f15a3a44c7768883b8fd6eb7d4f4c87adb8163b4. Actually, the real problem (to which I was thinking about years ago, and somehow I forgot) is that cloning is a dumb idea, as if `to` is less than the length of the array you are actually allocating useless space.

**Comment by Speiger** [comment]
@vigna i wouldn't have implemented CommonPool blindly like you did.
I would have provided an option to either use the common pool or not. Because a common pool is "Shared"
fastutils sorting algorythm relies on the fact that all its tasks are finished within the pool before it can continue.
It locks down the thread too.

Now the issue relies where what if someone has a multithreading process and fills the joined pool and has dependencies running with fastutil parallel sorting. That results in the threadpool maybe not shutting down once all fastutil tasks are done.
A dedicated threadpool is still needed.

TL:DR you might have just implemented a big mistake. the way you have done it at least.

**Comment by incaseoftrouble** [comment]
FYI: Java's `parallelStream()` has the same "problem" (it also always binds to the common pool)

**Comment by vigna** [comment]
Yes, the world is in danger, but @Speiger will come to the rescue 🤷🏻‍♂️.

**Comment by Speiger** [comment]
@vigna I haven't found out a solution that requires not 5 new functions to be created and that i am honestly happy with yet.
But the endresult will be most likely to have a custom class with a AtomicInteger for tasks started and tasks finished.

I figure something out. And if you want to clone it you are happy to do so.

**Comment by Speiger** [comment]
@vigna instead of shutting down the pool, the ForkJoinTask has a function called join or quietlyJoin. Which would fix the issue.
Also to address @incaseoftrouble after looking into parallelStream() no. It actually uses invoke/join on the task not the pool.
And this also has no issue with starting subtasks because only extra tasks (> 1) get their own thread allocated and the first task subtask gets executed on the same thread as the parent task.

This is 2 line fix, for the sorting functions, Except for Radix sort because that uses a different concept entirely and I haven't looked into that.
